### DISCUSSION TITLE
Internal updates and cleanup

### DISCUSSION THREAD

**Discussion by mloubout** [slimgroup/JUDI.jl#57] [General]
This is quite bigger than I anticipated when started this. Originally was supposed to be a small update to make `TimeProbeSeismic.jl` easier to overload. 

- Most of these changes are internal API updates and do not impact in any way the user interface and user code.  
- Some files are showing as "new" and "removed: while they are only moved, this is only de to directory reorganization and changes.
## Optimization

#43 
Removed `SLIM_Optim` that got moved to [SlimOptim.jl](https://github.com/slimgroup/SlimOptim.jl)

## judiJacobian

This is now reworked as an abstract type `judiAbstractJacobian` that makes "inheritance" easier. Now both judiJacobian and judiJacobianExtQ are subtypes of it make it a lot of duplicate code disappear.  Absolutely nothing changed on the user side so you don't need to care about it.

## typed implementation

I added and now enforce typed arguments for functions. It makes most of the codebase more juliaesque and therefore a tiny bit more efficient since it allows better dispatch on types. Once again nothing should have changed from the user perspective as there are still fallback definitions for input with not expected types.

## Parallelization

Switched the parallelization to `pmap` that makes it quite cleaner and added fallback cases for single source and/or serial julia to avoid the `pmap` overhead in these cases. In serial mode, it now uses `asyncmap` that may still have multiple tasks running at once.

## Misc

Few small fixes and updates:

- Fixed `limit_m` that seemed to have a shift somewhere and wasn't matching the data exactly. Fixes #55 
- Changed how python surce is loaded into `__init__` so that it is only loaded once and just used when needed.
- Small restructuring of the files to make navigation easier. Split into `Types` (judiVector, judiWeigths, ...), `Modeling` (modeling function such as fwi_objectiv), `LinearOperators` and `Utils`.
- Small cleaup of arithmetic (+, - , ..) for `Types`.
- Added shape information to shape error message for better debugging
- Added some docstring to auxiliary functions
- fix `return_array` with `dm=0` #51 
- Dimensional `cumsum` #47 #50

@ziyiyin97  This should fix your `limit_m` issues as well including your PR #54

<hr /><em>This discussion was created from the release <a href='https://github.com/slimgroup/JUDI.jl/releases/tag/v2.3.0'>Internal updates and cleanup</a>.</em>

**Comment by ziyiyin97** [comment]
Can you also add a note about `rebuild_jv` function here?

**Reply by mloubout** [reply]
Done
